src/factorizations/svd.jl

```diff
@@ -1,207 +1,239 @@
+
 """
-    SVD <: Factorization
-
-Matrix factorization type of the singular value decomposition (SVD) of a matrix `A`.
-This is the return type of [`svd(_)`](@ref), the corresponding matrix factorization function.
-
-If `F::SVD` is the factorization object, `U`, `S`, `V` and `Vt` can be obtained
-via `F.U`, `F.S`, `F.V` and `F.Vt`, such that `A = U * Diagonal(S) * Vt`.
-The singular values in `S` are sorted in descending order.
-
-Iterating the decomposition produces the components `U`, `S`, and `V`.
-
-# Examples
-```jldoctest
-julia> A = [1. 0. 0. 0. 2.; 0. 0. 3. 0. 0.; 0. 0. 0. 0. 0.; 0. 2. 0. 0. 0.]
-4×5 Matrix{Float64}:
- 1.0  0.0  0.0  0.0  2.0
- 0.0  0.0  3.0  0.0  0.0
- 0.0  0.0  0.0  0.0  0.0
- 0.0  2.0  0.0  0.0  0.0
-
-julia> F = BlockSparseArrays.svd(A)
-BlockSparseArrays.SVD{Float64, Float64, Matrix{Float64}, Vector{Float64}, Matrix{Float64}}
-U factor:
-4×4 Matrix{Float64}:
- 0.0  1.0   0.0  0.0
- 1.0  0.0   0.0  0.0
- 0.0  0.0   0.0  1.0
- 0.0  0.0  -1.0  0.0
-singular values:
-4-element Vector{Float64}:
- 3.0
- 2.23606797749979
- 2.0
- 0.0
-Vt factor:
-4×5 Matrix{Float64}:
- -0.0        0.0  1.0  -0.0  0.0
-  0.447214   0.0  0.0   0.0  0.894427
-  0.0       -1.0  0.0   0.0  0.0
-  0.0        0.0  0.0   1.0  0.0
-
-julia> F.U * Diagonal(F.S) * F.Vt
-4×5 Matrix{Float64}:
- 1.0  0.0  0.0  0.0  2.0
- 0.0  0.0  3.0  0.0  0.0
- 0.0  0.0  0.0  0.0  0.0
- 0.0  2.0  0.0  0.0  0.0
-
-julia> u, s, v = F; # destructuring via iteration
-
-julia> u == F.U && s == F.S && v == F.V
-true
-```
+    BlockDiagonalAlgorithm(A::MatrixAlgebraKit.AbstractAlgorithm)
+  
+A wrapper for `MatrixAlgebraKit.AbstractAlgorithm` that implements the wrapped algorithm on
+a block-by-block basis, which is possible if the input matrix is a block-diagonal matrix.
 """
-struct SVD{T,Tr,M<:AbstractArray{T},C<:AbstractVector{Tr},N<:AbstractArray{T}} <:
-       Factorization{T}
-  U::M
-  S::C
-  Vt::N
-  function SVD{T,Tr,M,C,N}(
-    U, S, Vt
-  ) where {T,Tr,M<:AbstractArray{T},C<:AbstractVector{Tr},N<:AbstractArray{T}}
-    Base.require_one_based_indexing(U, S, Vt)
-    return new{T,Tr,M,C,N}(U, S, Vt)
-  end
-end
-function SVD(U::AbstractArray{T}, S::AbstractVector{Tr}, Vt::AbstractArray{T}) where {T,Tr}
-  return SVD{T,Tr,typeof(U),typeof(S),typeof(Vt)}(U, S, Vt)
-end
-function SVD{T}(U::AbstractArray, S::AbstractVector{Tr}, Vt::AbstractArray) where {T,Tr}
-  return SVD(
-    convert(AbstractArray{T}, U),
-    convert(AbstractVector{Tr}, S),
-    convert(AbstractArray{T}, Vt),
-  )
+struct BlockDiagonalAlgorithm{A<:MatrixAlgebraKit.AbstractAlgorithm} <:
+       MatrixAlgebraKit.AbstractAlgorithm
+  alg::A
 end
 
-function SVD{T}(F::SVD) where {T}
-  return SVD(
-    convert(AbstractMatrix{T}, F.U),
-    convert(AbstractVector{real(T)}, F.S),
-    convert(AbstractMatrix{T}, F.Vt),
-  )
+function MatrixAlgebraKit.default_svd_algorithm(A::AbstractBlockSparseMatrix; kwargs...)
+  @assert blocktype(A) <: StridedMatrix{<:LinearAlgebra.BLAS.BlasFloat} "unsupported type:
```

Comment:
In general it is best to avoid using `@assert`, it is meant for catching things that in principle should never happen, rather than input mistakes. Also they have left it open that in future versions of Julia they might be removed at certain optimization levels (https://docs.julialang.org/en/v1/base/base/#Base.@assert). I think `ArgumentError` is more appropriate here.